### PR TITLE
eventPassCardFooter: add next intl provider for EventPassContractReve…

### PR DESCRIPTION
**Pull request by AlexandreG-tech**
## **User description**
…alContractButton and mock next/cache for test in updateEventPassCart

EventPasses.tsx and PasCardSelect.tsx : added NextIntlProvider

migration: change order_amount_handling to not count quantity on update

payment-admin: integration tests to make sure stripeCheckoutSession is deleted if the execution of the function goes right

updateEventPassCart: change migration to not count existing pendingOrder quantity in total quantity and added tests for those edge cases

cart/purchase : reason in search params

stripe process: add button and action to cancel a purchase order and a stripe session, added tests for the getStripeActiveCheckoutSession function and especially the no user/user mail case

lowercase for all eth addresses : update tests and add migration for the insert trigger ; make the cancel my purchase button reactive

___

## **Type**
Enhancement

___

## **Changes walkthrough**
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>17 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>nftActivity.integration.test.ts
</strong><dd><code>Update <code>fromAddress</code> and <code>toAddress</code> to use lowercase addresses in <br><code>nftActivity.integration.test.ts</code></code></dd></summary>
<hr>

libs/indexer/alchemy/webhooks/src/lib/nftActivity.integration.test.ts
<li>Updated the <code>fromAddress</code> and <code>toAddress</code> values in the <code>mockActivity</code> and <br><code>mockActivity2</code> objects to use lowercase addresses.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Offline-Project/marketplace/pull/271/files#diff-c098e6c6ba11ed0de016ab4337d7f5e533a5134bf9965b280f4ea64e8177b40c"></a></td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>packCollection.integration.test.ts
</strong><dd><code>Update <code>contractAddress</code> to use lowercase addresses in <br><code>packCollection.integration.test.ts</code></code></dd></summary>
<hr>

libs/nft/thirdweb-organizer/src/packCollection.integration.test.ts
<li>Updated the <code>contractAddress</code> values in the <code>mockActivity</code> and <br><code>mockActivity2</code> objects to use lowercase addresses.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Offline-Project/marketplace/pull/271/files#diff-548fbfa3ab145e3f1f4cb9a6a7b6fb8f863c4773a614ba058ace31bb2b889fc9"></a></td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>pendingOrder.spec.ts
</strong><dd><code>Add <code>eventPass3</code> and test case for updating order with new quantity in <br><code>pendingOrder.spec.ts</code></code></dd></summary>
<hr>

libs/gql/user/api/src/queries/cart/pendingOrder.spec.ts
<li>Added <code>eventPass3</code> object with a lowercase <code>contractAddress</code> value.<br> <li> Added a test case to check the update of an order with a new quantity <br>that exceeds the max limit.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Offline-Project/marketplace/pull/271/files#diff-bbdec54c1ef7348b3a49901320f7ea6c89954c2f928e3db1f6ad6305039236bf"></a></td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>payment-admin.integration.test.ts
</strong><dd><code>Add integration tests for <code>confirmedStripeCheckoutSession</code> in <br><code>payment-admin.integration.test.ts</code></code></dd></summary>
<hr>

libs/payment/admin/src/lib/payment-admin.integration.test.ts
<li>Added integration tests for the <code>confirmedStripeCheckoutSession</code> <br>function.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Offline-Project/marketplace/pull/271/files#diff-9abbd512a75f8ae7511d08a1ff32e0c7e0430dcf0401e49f4984f7a31b532fda"></a></td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>BoundedNumericStepper.tsx
</strong><dd><code>Refactor <code>BoundedNumericStepper</code> to use <code>value</code> prop and <code>onChange</code> callback <br>in <code>BoundedNumericStepper.tsx</code></code></dd></summary>
<hr>

libs/ui/components/src/lib/bounded-numeric-stepper/BoundedNumericStepper.tsx
<li>Removed the <code>useState</code> hook and replaced it with a <code>value</code> prop and <br><code>onChange</code> callback.<br> <li> Updated the <code>increment</code> and <code>decrement</code> functions to use the <code>value</code> prop <br>and <code>onChange</code> callback.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Offline-Project/marketplace/pull/271/files#diff-07fe8623ed00572e51cbe60efc025365d8227ca2292f3517f1a0dadf95ebfbe6"></a></td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>UserCart.tsx
</strong><dd><code>Add `reason` prop and section to `UserCart` in `UserCart.tsx`</code>&nbsp; &nbsp; </dd></summary>
<hr>

libs/features/cart/src/lib/UserCart/UserCart.tsx
<li>Added a new prop <code>reason</code> to <code>UserCartProps</code> and <code>UserCart</code> component.<br> <li> Added a new section in <code>UserCart</code> to display an alert if <code>reason</code> prop is <br>provided.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Offline-Project/marketplace/pull/271/files#diff-b7f73370b26d73e710d56b085d16d51e0727994905da20570c56e512bc1cf269"></a></td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>PassCardSelect.tsx
</strong><dd><code>Add `value` prop to `PassCardSelect` in `PassCardSelect.tsx`</code>&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

libs/features/organizer/event/src/lib/molecules/PassCard/PassCardSelect.tsx
<li>Added a new prop <code>value</code> to <code>PassCardSelectProps</code> and <br><code>PassCardSelectContent</code> component.<br> <li> Updated the <code>PassCardSelectContent</code> component to use the <code>value</code> prop.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Offline-Project/marketplace/pull/271/files#diff-f7eb4096cc252d96098d2f0345c3cc33c221eabcd39b773c94025ff609e7d509"></a></td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>PassPurchaseHeader.tsx
</strong><dd><code>Add <code>CancelPurchaseButton</code> component to <code>PassPurchaseHeader</code> in <br><code>PassPurchaseHeader.tsx</code></code></dd></summary>
<hr>

libs/features/organizer/event/src/lib/molecules/PassPurchaseHeader/PassPurchaseHeader.tsx
<li>Added a new component <code>CancelPurchaseButton</code> to handle canceling a <br>purchase order.<br> <li> Updated the <code>PurchaseInProgressAlert</code> component to include the <br><code>CancelPurchaseButton</code>.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Offline-Project/marketplace/pull/271/files#diff-38096d483193006b2546d0eb5279724cd43f98ed9b478f83a6dbf9c905335671"></a></td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>CancelPurchaseButtonClient.tsx
</strong><dd><code>Add <code>CancelPurchaseButton</code> component to handle canceling a purchase <br>order in <code>CancelPurchaseButtonClient.tsx</code></code></dd></summary>
<hr>

libs/features/organizer/event/src/lib/molecules/PassPurchaseHeader/CancelPurchaseButtonClient.tsx
<li>Added a new component <code>CancelPurchaseButton</code> to handle canceling a <br>purchase order.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Offline-Project/marketplace/pull/271/files#diff-3dd1fa2ea93cce4f8a017510d7eebd47a1513554e334321f7e1c6ab334aa6257"></a></td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>CartCancelled.stories.tsx
</strong><dd><code>Add story for <code>CartCancelled</code> component with cancel purchase button in <br><code>CartCancelled.stories.tsx</code></code></dd></summary>
<hr>

libs/features/cart/src/lib/CartCancelled/CartCancelled.stories.tsx
<li>Added a new story for <code>CartCancelled</code> component with cancel purchase <br>button.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Offline-Project/marketplace/pull/271/files#diff-c0839eacbbaae356cc0850ee45ca9edc4ffcdbcf3a31567153fed55640370b57"></a></td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>action.ts
</strong><dd><code>Add <code>cancelPurchaseForUser</code> function to handle canceling a purchase <br>order in <code>action.ts</code></code></dd></summary>
<hr>

libs/features/organizer/event/src/lib/molecules/PassPurchaseHeader/action.ts
<li>Added a new function <code>cancelPurchaseForUser</code> to handle canceling a <br>purchase order.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Offline-Project/marketplace/pull/271/files#diff-7b5cbc2e828732412025c534f6b9ebb1f433022a02d62d6ece95c4cb7fc619d0"></a></td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>getStripeActiveCheckoutSession.ts
</strong><dd><code>Update <code>getStripeActiveCheckoutSession</code> to throw error if user has no <br>email in <code>getStripeActiveCheckoutSession.ts</code></code></dd></summary>
<hr>

libs/features/payment-api/src/lib/getStripeActiveCheckoutSession.ts
<li>Updated the <code>getStripeActiveCheckoutSession</code> function to throw an error <br>if the user has no email.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Offline-Project/marketplace/pull/271/files#diff-d50c534e3a789f2a439302d0cb0bb5423fea22374e6b9b44b60c21657ae60811"></a></td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>updateEventPassCart.ts
</strong><dd><code>Update <code>updateEventPassCart</code> to revalidate tags in <br><code>updateEventPassCart.ts</code></code></dd></summary>
<hr>

libs/features/organizer/event-actions/src/lib/updateEventPassCart.ts
<li>Updated the <code>updateEventPassCart</code> function to revalidate tags even when <br>running in Jest.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Offline-Project/marketplace/pull/271/files#diff-42a1df0e01efc1a53f3e15b0ece1f652e67c27ca5dee6b8e5a1b86b892e473b6"></a></td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>PassPurchase.stories.tsx
</strong><dd><code>Update <code>WithPurchaseInProcess</code> story to include cancel purchase button <br>in <code>PassPurchase.stories.tsx</code></code></dd></summary>
<hr>

libs/features/organizer/event/src/lib/pages/PassPurchase/PassPurchase.stories.tsx
<li>Updated the <code>WithPurchaseInProcess</code> story to include the cancel purchase <br>button.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Offline-Project/marketplace/pull/271/files#diff-bd0eb014551960eda79f3c8d929400405355b226ddec9ce4501f6f91dc732e59"></a></td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>EventPassNftFilesTableClient.tsx
</strong><dd><code>Return <code>null</code> if <code>data</code> array is empty in <code>EventPassNftFilesTableClient.tsx</code></code></dd></summary>
<hr>

libs/features/back-office/events/src/lib/molecules/EventPassNftFilesTable/EventPassNftFilesTableClient.tsx
- Added a check to return `null` if the `data` array is empty.

</details>
    

  </td>
  <td><a href="https://github.com/Offline-Project/marketplace/pull/271/files#diff-b29358cf35e279788aa8d21626b8ed1bdd97c64bdd1b258f1dc52019e2f2695b"></a></td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>EventPassCardFooter.tsx
</strong><dd><code>Replace `_locale` with `locale` in `EventPassCardFooter.tsx`</code>&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

libs/features/back-office/events/src/lib/molecules/EventPassCardFooter/EventPassCardFooter.tsx
<li>Removed the <code>_locale</code> variable and replaced it with <code>locale</code> variable.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Offline-Project/marketplace/pull/271/files#diff-2a8396595ed3f28672c103492680b0c3fe66186565cf7e97856170be0c0b3444"></a></td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>getStripeActiveCheckoutSession.spec.ts
</strong><dd><code>Add unit tests for <code>getStripeActiveCheckoutSession</code> in <br><code>getStripeActiveCheckoutSession.spec.ts</code></code></dd></summary>
<hr>

libs/features/payment-api/src/lib/getStripeActiveCheckoutSession.spec.ts
<li>Added unit tests for the <code>getStripeActiveCheckoutSession</code> function.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Offline-Project/marketplace/pull/271/files#diff-154014defc828f71805a50f73a66697af699615ac4e5342ae52772efa5519bab"></a></td>
</tr>                    
</table></details></td></tr></tr></tbody></table><hr>

<details> <summary><strong>✨ Usage guide:</strong></summary><hr> 

**Overview:**
The `describe` tool scans the PR code changes, and generates a description for the PR - title, type, summary, walkthrough and labels. The tool can be triggered [automatically](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#github-app-automatic-tools) every time a new PR is opened, or can be invoked manually by commenting on a PR.

When commenting, to edit [configurations](https://github.com/Codium-ai/pr-agent/blob/main/pr_agent/settings/configuration.toml#L46) related to the describe tool (`pr_description` section), use the following template:
```
/describe --pr_description.some_config1=... --pr_description.some_config2=...
```
With a [configuration file](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#working-with-github-app), use the following template:
```
[pr_description]
some_config1=...
some_config2=...
```

<table><tr><td><details> <summary><strong> Enabling\disabling automation </strong></summary><hr>

- When you first install the app, the [default mode](https://github.com/Codium-ai/pr-agent/blob/main/Usage.md#github-app-automatic-tools) for the describe tool is:
```
pr_commands = ["/describe --pr_description.add_original_user_description=true" 
                         "--pr_description.keep_original_user_title=true", ...]
```
meaning the `describe` tool will run automatically on every PR, will keep the original title, and will add the original user description above the generated description. 

- Markers are an alternative way to control the generated description, to give maximal control to the user. If you set:
```
pr_commands = ["/describe --pr_description.use_description_markers=true", ...]
```
the tool will replace every marker of the form `pr_agent:marker_name` in the PR description with the relevant content, where `marker_name` is one of the following:
  - `type`: the PR type.
  - `summary`: the PR summary.
  - `walkthrough`: the PR walkthrough.

Note that when markers are enabled, if the original PR description does not contain any markers, the tool will not alter the description at all.
        

</details></td></tr>

<tr><td><details> <summary><strong> Custom labels </strong></summary><hr>

The default labels of the `describe` tool are quite generic: [`Bug fix`, `Tests`, `Enhancement`, `Documentation`, `Other`].

If you specify [custom labels](https://github.com/Codium-ai/pr-agent/blob/main/docs/DESCRIBE.md#handle-custom-labels-from-the-repos-labels-page-gem) in the repo's labels page or via configuration file, you can get tailored labels for your use cases.
Examples for custom labels:
- `Main topic:performance` - pr_agent:The main topic of this PR is performance
- `New endpoint` - pr_agent:A new endpoint was added in this PR
- `SQL query` - pr_agent:A new SQL query was added in this PR
- `Dockerfile changes` - pr_agent:The PR contains changes in the Dockerfile
- ...

The list above is eclectic, and aims to give an idea of different possibilities. Define custom labels that are relevant for your repo and use cases.
Note that Labels are not mutually exclusive, so you can add multiple label categories.
Make sure to provide proper title, and a detailed and well-phrased description for each label, so the tool will know when to suggest it.        

</details></td></tr>

<tr><td><details> <summary><strong> Inline File Walkthrough 💎</strong></summary><hr>

For enhanced user experience, the `describe` tool can add file summaries directly to the "Files changed" tab in the PR page.
This will enable you to quickly understand the changes in each file, while reviewing the code changes (diffs).

To enable inline file summary, set `pr_description.inline_file_summary` in the configuration file, possible values are:
- `'table'`: File changes walkthrough table will be displayed on the top of the "Files changed" tab, in addition to the "Conversation" tab.
- `true`: A collapsable file comment with changes title and a changes summary for each file in the PR.
- `false` (default): File changes walkthrough will be added only to the "Conversation" tab.
<tr><td><details> <summary><strong> Utilizing extra instructions</strong></summary><hr>

The `describe` tool can be configured with extra instructions, to guide the model to a feedback tailored to the needs of your project.

Be specific, clear, and concise in the instructions. With extra instructions, you are the prompter. Notice that the general structure of the description is fixed, and cannot be changed. Extra instructions can change the content or style of each sub-section of the PR description.

Examples for extra instructions:
```
[pr_description] 
extra_instructions="""
- The PR title should be in the format: '<PR type>: <title>'
- The title should be short and concise (up to 10 words)
- ...
"""
```
Use triple quotes to write multi-line instructions. Use bullet points to make the instructions more readable.

</details></td></tr>

<tr><td><details> <summary><strong> More PR-Agent commands</strong></summary><hr> 

> To invoke the PR-Agent, add a comment using one of the following commands:  
> - **/review**: Request a review of your Pull Request.   
> - **/describe**: Update the PR title and description based on the contents of the PR.   
> - **/improve [--extended]**: Suggest code improvements. Extended mode provides a higher quality feedback.   
> - **/ask \<QUESTION\>**: Ask a question about the PR.   
> - **/update_changelog**: Update the changelog based on the PR's contents.   
> - **/add_docs** 💎: Generate docstring for new components introduced in the PR.   
> - **/generate_labels** 💎: Generate labels for the PR based on the PR's contents.   
> - **/analyze** 💎: Automatically analyzes the PR, and presents changes walkthrough for each component.   

>See the [tools guide](https://github.com/Codium-ai/pr-agent/blob/main/docs/TOOLS_GUIDE.md) for more details.
>To list the possible configuration parameters, add a **/config** comment.   
 

</details></td></tr>

</table>

See the [describe usage](https://github.com/Codium-ai/pr-agent/blob/main/docs/DESCRIBE.md) page for a comprehensive guide on using this tool.

</details>

### CODE REVIEW

File: hasura/app/migrations/default/1707365734019_insert_lowercase_trigger/up.sql [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+$$ LANGUAGE plpgsql;
+
+CREATE TRIGGER nftTransfer_before_insert
+BEFORE INSERT ON "nftTransfer"
+FOR EACH ROW EXECUTE FUNCTION force_lowercase_nftTransfer();
+
+-- Function for eventPassNft table
+CREATE OR REPLACE FUNCTION force_lowercase_eventPassNft() RETURNS TRIGGER AS $$
+BEGIN
+    NEW."contractAddress" := LOWER(NEW."contractAddress");
```
Comment (sebpalluel) [inline hasura/app/migrations/default/1707365734019_insert_lowercase_trigger/up.sql]: In this case we need to apply the lowercase also on update for currentOwnerAddress because it's updated.
+ Same applies for `packNftSupply` that is missing.
+ add trigger for `packNftContract`

File: libs/features/back-office/events/src/lib/molecules/EventPassFilesUploader/EventPassFilesUploader.tsx [hunk 2]
```diff
@@ -27,7 +29,7 @@ export interface EventPassFilesTableProps
   buttonClassName?: string;
 }
 
-export function EventPassFilesUploader(props: EventPassFilesTableProps) {
+export async function EventPassFilesUploader(props: EventPassFilesTableProps) {
```
Comment (sebpalluel) [inline libs/features/back-office/events/src/lib/molecules/EventPassFilesUploader/EventPassFilesUploader.tsx:32]: Not necessary, Suspense handle the async from the children `EventPassFilesUploaderContent` that is indeed asynchronous.

File: libs/features/organizer/event/src/lib/commons/action.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
+'use server';
```
Comment (sebpalluel) [inline libs/features/organizer/event/src/lib/commons/action.ts:1]: Add unit test for this one

File: libs/features/organizer/event/src/lib/molecules/PassCard/PassCardSelectClient.tsx [hunk 4]
```diff
@@ -5,31 +5,37 @@ import {
...
   const [isPending, startTransition] = useTransition();
+  const router = useRouter();
+  const [value, setValue] = useState(
+    boundedNumericStepperProps.initialValue || 0,
+  );
   const { toast } = useToast();
   const t = useTranslations('Organizer.Event.PassPurchase');
 
   return (
     <BoundedNumericStepper
```
Comment (sebpalluel) [inline libs/features/organizer/event/src/lib/molecules/PassCard/PassCardSelectClient.tsx:36]: Would be better to use a revalidateTag on the action (to avoid weird UX on the Sheet and intercepted route so pass purchase card version), instead of router.refresh, if nextjs is not buggy as it used to with this

**Review by sebpalluel — CHANGES_REQUESTED** [review]

File: libs/features/organizer/event/src/lib/molecules/PassCard/PassCardSelectClient.tsx (hunk 4 above)
Comment (AlexandreG-tech) [reply libs/features/organizer/event/src/lib/molecules/PassCard/PassCardSelectClient.tsx:36]: Without the router.refresh on the first completion of the onChange function, it seems to switch to the /purchase instead of staying on the sheet .purchase, could be a revalidateTag causing a reload and eventually loading the /purchase because of the URL

File: libs/features/back-office/events/src/lib/molecules/EventPassFilesUploader/EventPassFilesUploader.tsx [hunk 5]
```diff
@@ -29,7 +29,7 @@ export interface EventPassFilesTableProps
...
```
Comment (sebpalluel) [inline libs/features/back-office/events/src/lib/molecules/EventPassFilesUploader/EventPassFilesUploader.tsx:32]: Well no the issue here really is the `'use server';` at the top of the file that should be removed.

File: hasura/app/migrations/default/1707365734019_squashed/up.sql [hunk 6]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (sebpalluel) [inline hasura/app/migrations/default/1707365734019_squashed/up.sql]: We need to give a name after the unix timestamp that say shortly what's the migration about

**Review by sebpalluel — CHANGES_REQUESTED** [review]

File: libs/features/back-office/events/src/lib/molecules/EventPassFilesUploader/EventPassFilesUploader.tsx (hunk 5 above)
Comment (sebpalluel) [reply libs/features/back-office/events/src/lib/molecules/EventPassFilesUploader/EventPassFilesUploader.tsx:32]: That way in the client context it's going to be a client component and on the server side a RSC. If it works ?

Comment (sebpalluel) [reply libs/features/back-office/events/src/lib/molecules/EventPassFilesUploader/EventPassFilesUploader.tsx:32]: Oh no you are right, we are using `getTranslations` from server so never mind

**Review by sebpalluel — APPROVED** [review]
